### PR TITLE
Implement elastic reasoning

### PR DESCRIPTION
**Pull request by mikasenghaas**
This PR implements configuring a "thinking" and "solution" budget as defined in the paper Scalable Chain of Thoughts via Elastic Reasoning ([arXiv](https://arxiv.org/pdf/2505.05315)). The main idea is that we want to force a model to produce a solution even if the CoT is unfinished, with the hope being that an unfinished thinking trace still has value.

In the original paper the authors propose appending a `</think>` token if the thinking budget is reached. In their official code ([link](https://github.com/SalesforceAIResearch/Elastic-Reasoning/blob/8dacccd1ff60e82caa9033844cb0f07b97884259/verl/verl/workers/rollout/vllm_rollout/vllm_rollout.py#L303)) they do two separate calls to the inference engine, one with the think budget and one with the solution budget. If the last sampled token in the first stage is not `</think` they append it manually. In our implementation we use a post-hook on the sampler which replaces the token sampled when reaching the thinking budget by `</think`. This is more elegant because we do not require any other code changes to our inference script - we still do a single call to `.generate()` during an inference step.

One can configure the elastic reasoning by setting the `--sampling.max-solution-tokens` configuration which will dyncamilly compute the think budget as `sampling.max-tokens - sampling.max-solution-tokens`. We chose this API because we can set a constant default value for the solution budget and scale the thinking budget with the context length of the model

## Examples

1. No thinking budget set (default)

```bash
uv run src/zeroband/infer.py @ configs/inference/debug.toml --sampling.max-tokens 10 --log-level debug
```

This produces the following output

```txt
<|im_start|>assistant
<think>
Okay, let me try to solve this
```

2. Thinking budget is set

We dynamically compute the thinking budget as the difference of the maximum number of output tokens (this is `min(context length - num input tokens, max output tokens)`). In this example, the context length is 40k by default and we set the max number of output tokens to 10, so here we will almost always hit the thinking budget after 5 output tokens

```bash
uv run src/zeroband/infer.py @ configs/inference/debug.toml --sampling.max-tokens 10 --sampling.max-solution-tokens 5 --log-level debug
```

This produces the following output

```txt
<|im_start|>assistant
<think>
Okay,</think>

We are given a
```

In the second example, we do not set `--sampling.max-tokens`and so the thinking budget is dynamically computed as the difference between the context length (2k)  and the number of input tokens for each sequence, leading to a dynamic thinking budget.

```bash
uv run src/zeroband/infer.py @ configs/inference/debug.toml --model.max-model-len 2048 --sampling.max-tokens None  --sampling.max-solution-tokens 5 --log-level debug
```

### CODE REVIEW

File: src/zeroband/inference/config.py [hunk 1]
```diff
@@ -280,6 +281,9 @@ class Config(BaseSettings):
     # The RL configuration. If None, inference will run in a non-RL setting.
     rl: Annotated[RLConfig | None, Field(default=None)]
 
+    # The elastic reasoning configuration
+    elastic_reasoning: Annotated[ElasticReasoningConfig, Field(default=ElasticReasoningConfig())]
```
Comment (samsja) [inline src/zeroband/inference/config.py:285]:
```suggestion
    elastic_reasoning: ElasticReasoningConfig,  =  ElasticReasoningConfig() 
 
```

isn't better to keep like this ?

File: src/zeroband/inference/elastic_reasoning.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+from vllm import LLM
+from vllm.model_executor.layers.sampler import SamplerOutput, SamplingMetadata
+
+from zeroband.utils.config import BaseConfig
+from zeroband.utils.logger import get_logger
+
+
+class ElasticReasoningConfig(BaseConfig):
+    """Configures elastic reasoning from `Scalable Chain of Thoughts via Elastic Reasoning` (https://arxiv.org/abs/2505.05315)."""
+
```
Comment (samsja) [inline src/zeroband/inference/elastic_reasoning.py]: lets keep config on the config.py of inference

File: src/zeroband/inference/elastic_reasoning.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    enable: Annotated[bool, Field(default=False, description="Whether to enable elastic reasoning.")]
+    think_budget: Annotated[
+        int,
+        Field(
+            default=1024,
```
Comment (samsja) [inline src/zeroband/inference/elastic_reasoning.py]: I like the default being pass by the = rather that defualt

File: src/zeroband/inference/elastic_reasoning.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    solution_budget: Annotated[int, Field(default=1024, description="Number of tokens to allow for solution.")]
+    stop_think_token_id: Annotated[
+        int,
+        Field(
+            default=151668, description="The token ID of the `</think>` token which is swapped in to force the model to produce a solution."
+        ),
+    ]  # </think
+
+
+def swap_think_token(_, __, kwargs, outputs, config: ElasticReasoningConfig) -> SamplerOutput | None:
```
Comment (samsja) [inline src/zeroband/inference/elastic_reasoning.py]:
```suggestion
def swap_think_token(_, __, kwargs, outputs, config: ElasticReasoningConfig) -> SamplerOutput | None:
```

what is +_ and __ doing ?

File: src/zeroband/inference/elastic_reasoning.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        for seq_group, seq_outputs in zip(sampling_metadata.seq_groups, sampling_output.outputs):
+            for seq_id, seq_data, seq_output in zip(seq_group.seq_data.keys(), seq_group.seq_data.values(), seq_outputs.samples):
+                num_output_tokens = len(seq_data.output_token_ids) + 1  # Increment by 1 because the current token is not yet in the output
+                if num_output_tokens == config.think_budget:
+                    get_logger("INFER").debug(
+                        f"Thinking budget reached for sequence {seq_id}. Replacing sampled token {seq_output.output_token} with {config.stop_think_token_id}"
+                    )
+                    seq_output.logprobs = {config.stop_think_token_id: seq_output.logprobs[seq_output.output_token]}
+                    seq_output.output_token = config.stop_think_token_id
+
```
Comment (samsja) [inline src/zeroband/inference/elastic_reasoning.py]: we need to have way to flag that this token was generated manually. 

We need to mask it for the loss

File: src/zeroband/infer.py [hunk 6]
```diff
@@ -130,6 +131,10 @@ def inference(config: InferenceConfig):
     logger.info(f"Initializing sampling parameters ({config.sampling} seed={config.seed})")
     sampling_params = SamplingParams(**config.sampling.model_dump(), seed=config.seed)
 
+    # Setup elastic reasoning
+    # Note: This hook has to be set up before the PP communication hooks to ensure that possibly modified sampler outputs are propagated to group peers
+    setup_elastic_reasoning(config.elastic_reasoning, llm=llm)
```
Comment (samsja) [inline src/zeroband/infer.py]:
```suggestion
    if config.elastic_reasoning is not None:
            setup_elastic_reasoning(config.elastic_reasoning, llm=llm)
```

File: src/zeroband/inference/elastic_reasoning.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    return sampling_output
+
+
+def setup_elastic_reasoning(config: ElasticReasoningConfig, llm: LLM):
+    # Skip if disabled
+    if not config.enable:
+        return
```
Comment (samsja) [inline src/zeroband/inference/elastic_reasoning.py]:
```suggestion
```

File: src/zeroband/inference/elastic_reasoning.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+from vllm.model_executor.layers.sampler import SamplerOutput, SamplingMetadata
+
+from zeroband.utils.config import BaseConfig
+from zeroband.utils.logger import get_logger
+
+
+class ElasticReasoningConfig(BaseConfig):
+    """Configures elastic reasoning from `Scalable Chain of Thoughts via Elastic Reasoning` (https://arxiv.org/abs/2505.05315)."""
+
+    enable: Annotated[bool, Field(default=False, description="Whether to enable elastic reasoning.")]
```
Comment (samsja) [inline src/zeroband/inference/elastic_reasoning.py]:
```suggestion
```

File: src/zeroband/inference/config.py (hunk 1 above)
Comment (samsja) [inline src/zeroband/inference/config.py]:
```suggestion
    elastic_reasoning: ElasticReasoningConfig | None = None
```

**Review by samsja — COMMENTED** [review]
code looks good, I added some stuff with respect to config. 

lets rather use the X | None pattern instead of having enable flag

File: src/zeroband/inference/elastic_reasoning.py (hunk 2 above)
Comment (mikasenghaas) [reply src/zeroband/inference/elastic_reasoning.py]: oh, okay. i was actually step by step moving subconfigs into submodules whenver possible (see rewards and pipeline). if we decide to do this, then i would move back all the configs to be consistent. executive decision given?

File: src/zeroband/inference/elastic_reasoning.py (hunk 3 above)
Comment (mikasenghaas) [reply src/zeroband/inference/elastic_reasoning.py]: we pass default for all configs now so this would be a changed across all config values. 

the main reason i decided for this is because i think it's more readable (less line breaks) and because if we ever want to dynamically set a default param we can do so by setting `default_factory: Callable`, e.g. `id: Field(default_factory: lambda: uuid.hex)` which is closer in syntax to `default` than the `=` way

File: src/zeroband/inference/elastic_reasoning.py (hunk 4 above)
Comment (mikasenghaas) [reply src/zeroband/inference/elastic_reasoning.py]: it's capturing the nn module that is being called (`_`) and the args (`__`) which are both not used but are required for the hook to be installed correctly

File: src/zeroband/inference/elastic_reasoning.py (hunk 5 above)
Comment (mikasenghaas) [reply src/zeroband/inference/elastic_reasoning.py]: mmh, we should only need to do this if it causes any issues. my guess would be that the effect on the overall traininig dynamic to be quite minimal, but will check with some more tests

not super clear to me how to propagate that "masking" information out of the hook

File: src/zeroband/infer.py (hunk 6 above)
Comment (mikasenghaas) [reply src/zeroband/infer.py]: yep will change to that

File: src/zeroband/inference/elastic_reasoning.py (hunk 2 above)
Comment (mikasenghaas) [reply src/zeroband/inference/elastic_reasoning.py]: moving the subconfigs for this pr into `src/zeroband/inference/config.py` and will move the rest in separate pr

File: src/zeroband/inference/elastic_reasoning.py (hunk 3 above)
Comment (samsja) [reply src/zeroband/inference/elastic_reasoning.py]: i see makes sense

File: src/zeroband/inference/elastic_reasoning.py (hunk 5 above)
Comment (samsja) [reply src/zeroband/inference/elastic_reasoning.py]: defo hard requirement here imo. Its quite clear that this would create issue during rl training because the logprob would be super wrong. Also, this is standard with things like tool call

**Comment by justusmattern27** [comment]
re discussion: </think> tokens should definitely be masked imo - if we train the model on positive advantage samples with the token unmasked, we specifically encourage the model to interrupt its thinking which will degrade performance; we want to encourage the model to think for longer, and having a bunch of training data in there of short sequence lengths will most likely cause the model to interrupt thinking after short amounts of time
